### PR TITLE
feat(colors): accept more formats

### PR DESCRIPTION
**Pull request by uncenter**
why not lol

### CODE REVIEW

File: src/commands/utils/color.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import { AttachmentBuilder, EmbedBuilder } from "discord.js";
 import sharp from "sharp";
 import { SlashCommand } from "../_types";
+// @ts-expect-error CJS
```
Comment (ryanccn) [inline src/commands/utils/color.ts:4]: no

File: src/commands/utils/color.ts [hunk 2]
```diff
@@ -16,6 +18,9 @@ export const colorCommand: SlashCommand = async (i) => {
 	} else if (hexRegex.test("#" + input)) {
 		fullString = "#" + input;
 		hexOnly = input;
+	} else if (Object.prototype.hasOwnProperty.call(cssColorKeywords as { [key: string]: string }, input)) {
```
Comment (ryanccn) [inline src/commands/utils/color.ts]: NO

File: src/commands/utils/color.ts (hunk 1 above)
Comment (uncenter) [reply src/commands/utils/color.ts:4]: > no

do you want a module declaration or what

File: src/commands/utils/color.ts (hunk 2 above)
Comment (uncenter) [reply src/commands/utils/color.ts]: > NO

whats the issue here, also i updated it a bit

Comment (ryanccn) [reply src/commands/utils/color.ts]: Does `input in cssColorKeywords` not work

File: src/commands/utils/color.ts (hunk 1 above)
Comment (ryanccn) [reply src/commands/utils/color.ts:4]: Use a better package or just vendorize the whole thing

Comment (uncenter) [reply src/commands/utils/color.ts:4]: > Use a better package or just vendorize the whole thing

... wdym by the second one

File: src/commands/utils/color.ts (hunk 2 above)
Comment (uncenter) [reply src/commands/utils/color.ts]: > Does `input in cssColorKeywords` not work

oops

File: src/commands/utils/color.ts (hunk 1 above)
Comment (uncenter) [reply src/commands/utils/color.ts:4]: all the packages are from before ESM so ig i could make my own but its just one comment

Comment (ryanccn) [reply src/commands/utils/color.ts:4]: Copy the colors into this codebase, from the package or wherever

Comment (uncenter) [reply src/commands/utils/color.ts:4]: > Copy the colors into this codebase, from the package or wherever

i was going to do that but where do you want me to put them

Comment (ryanccn) [reply src/commands/utils/color.ts:4]: `lib` probably

File: src/commands/utils/color.ts [hunk 3]
```diff
@@ -1,8 +1,16 @@
...
 import sharp from "sharp";
 import { SlashCommand } from "../_types";
-// @ts-expect-error CJS
-import cssColorKeywords from "css-color-keywords";
+import namedColors from "../../lib/namedColors.json" assert { type: "json" };
+
+function getHexCodeForColor(colorName: string) {
+	colorName = colorName.toLowerCase();
+	for (const hexCode in namedColors) {
+		const colorNames = (namedColors as { [key: string]: string[] })[hexCode].map((name) => name.toLowerCase());
```
Comment (uncenter) [inline src/commands/utils/color.ts:9]: not sure where i should be annotating `namedColors`'s type...

File: src/commands/utils/color.ts [hunk 4]
```diff
@@ -12,6 +12,14 @@ function getHexCodeForColor(colorName: string) {
 	return null;
 }
 
+function getNamesForHexCode(hexCode: string) {
+	hexCode = hexCode.toUpperCase();
+	if (hexCode in namedColors) {
+		return (namedColors as { [key: string]: string[] })[hexCode];
```
Comment (uncenter) [inline src/commands/utils/color.ts:18]: i hate this please help where should i type it properly

Comment (ryanccn) [reply src/commands/utils/color.ts:18]: Turn the JSON file into a module

**Comment by uncenter** [comment]
🥺

**Comment by uncenter** [comment]
Switching to https://github.com/elchininet/ColorTranslator lets us use any of the following string inputs:

| Example of CSS string inputs          | Description                                                                                                        |
| ------------------------------------- | ------------------------------------------------------------------------------------------------------------------ |
| `fuchsia`                             | Color keyword                                                                                                      |
| `#FF00FF`                             | Hexadecimal color                                                                                                  |
| `#F0F`                                | Shorthand hexadecimal color                                                                                        |
| `#FF00FF80`                           | Hexadecimal color with alpha                                                                                       |
| `#F0FF`                               | Shorthand hexadecimal color with alpha                                                                             |
| `rgb(255 0 255)`                      | Functional RGB notation                                                                                            |
| `rgb(100% 0% 100%)`                   | Functional RGB notation with values in percentages                                                                 |
| `rgb(255 0 255 / 0.5)`                | Functional RGB notation with alpha                                                                                 |
| `rgb(255 0 255 / 50%)`                | Functional RGB notation CSS with alpha in percenatages                                                             |
| `rgb(100% 0% 100% / 0.5)`             | Functional RGB notation with values in percentages and alpha                                                       |
| `rgb(100% 0% 100% / 50%)`             | Functional RGB notation with values and alpha in percentages                                                       |
| `rgb(255, 0, 255)`                    | Functional RGB notation (CSS Colors 3 comma-separated)                                                             |
| `rgba(255, 0, 255, 0.5)`              | Functional RGB notation with alpha (CSS Colors 3 comma-separated)                                                  |
| `hsl(300 100% 50%)`                   | Functional HSL notation                                                                                            |
| `hsl(300deg 100% 50%)`                | Functional HSL notation with hue in degrees                                                                        |
| `hsl(270grad 100% 50%)`               | Functional HSL notation with hue in gradians                                                                       |
| `hsl(5.24rad 100% 50%)`               | Functional HSL notation with hue in radians                                                                        |
| `hsl(0.83turn 100% 50%)`              | Functional HSL notation with hue in turns                                                                          |
| `hsl(300 100% 50% / 0.5)`             | Functional HSL notation with alpha                                                                                 |
| `hsl(300 100% 50% / 50%)`             | Functional HSL notation with alpha in percentages                                                                  |
| `hsl(300, 100%, 50%)`                 | Functional HSL notation (Colors 3 comma-separated)                                                                 |
| `hsl(300deg, 100%, 50%)`              | Functional HSL notation with hue in degrees (CSS Colors 3 comma-separated)                                         |
| `hsl(270grad, 100%, 50%)`             | Functional HSL notation with hue in gradians (CSS Colors 3 comma-separated)                                        |
| `hsl(5.24rad, 100%, 50%)`             | Functional HSL notation with hue in radians (CSS Colors 3 comma-separated)                                         |
| `hsl(0.83turn, 100%, 50%)`            | Functional HSL notation with hue in turns (CSS Colors 3 comma-separated)                                           |
| `hsla(300, 100%, 50%, 0.5)`           | Functional HSL notation with alpha (CSS Colors 3 comma-separated)                                                  |
| `device-cmyk(0% 100% 100% 0%)`        | Device-dependent functional CMYK notation with percentages                                                         |
| `device-cmyk(0% 100% 100% 0% / 1)`    | Device-dependent functional CMYK notation with percentages and alpha                                               |
| `device-cmyk(0% 100% 100% 0% / 100%)` | Device-dependent functional CMYK notation with percentages and alpha in percentages                                |
| `device-cmyk(0%, 100%, 100%, 0%)`     | Device-dependent functional CMYK notation with percentages (CSS Colors 3 comma-separated)                          |
| `device-cmyk(0%, 100%, 100%, 0%, 1)`  | Device-dependent functional CMYK notation with percentages and alpha (CSS Colors 3 comma-separated)                |
| `device-cmyk(0 1 1 0)`                | Device-dependent functional CMYK notation with numbers                                                             |
| `device-cmyk(0 1 1 0 / 1)`            | Device-dependent functional CMYK notation with numbers and alpha                                                   |
| `device-cmyk(0 1 1 0 / 100%)`         | Device-dependent functional CMYK notation with numbers and alpha in percentages                                    |
| `device-cmyk(0, 1, 1, 0)`             | Device-dependent functional CMYK notation with numbers (CSS Colors 3 comma-separated)                              |
| `device-cmyk(0, 1, 1, 0, 1)`          | Device-dependent functional CMYK notation with numbers and alpha (CSS Colors 3 comma-separated)                    |
| `cmyk(0% 100% 100% 0%)`               | Functional CMYK notation with percentages                                                                          |
| `cmyk(0% 100% 100% 0% / 1)`           | Functional CMYK notation with percentages and alpha                                                                |
| `cmyk(0% 100% 100% 0% / 100%)`        | Functional CMYK notation with percentages and alpha in percentages                                                 |
| `cmyk(0 1 1 0)`                       | Functional CMYK notation with numbers                                                                              |
| `cmyk(0 1 1 0 / 1)`                   | Functional CMYK notation with numbers and alpha                                                                    |
| `cmyk(0 1 1 0 / 100%)`                | Functional CMYK notation with numbers and alpha in percentages                                                     |
| `cmyk(0%, 100%, 100%, 0%)`            | Functional CMYK notation with percentages (CSS Colors 3 comma-separated)                                           |
| `cmyk(0%, 100%, 100%, 0%, 1)`         | Functional CMYK notation with percentages and alpha (CSS Colors 3 comma-separated)                                 |
| `cmyk(0, 1, 1, 0)`                    | Functional CMYK notation with numbers (CSS Colors 3 comma-separated)                                               |
| `cmyk(0, 1, 1, 0, 1)`                 | Functional CMYK notation with numbers and alpha (CSS Colors 3 comma-separated)                                     |

And we don't have to "normalize" anything because it just spits all of those formats right back out. The library seems well-tested and I did a bunch of my own tests with it and it works perfectly. I'm surprised how little it is used, but there aren't any active issues and it has recently seen a major update including types, the latest CSS color keywords, etc (https://github.com/elchininet/ColorTranslator/releases/tag/v3.0.2, 4 days ago).

File: src/commands/utils/color.ts [hunk 5]
```diff
@@ -3,19 +3,30 @@ import sharp from "sharp";
 import { SlashCommand } from "../_types";
 import { ColorTranslator } from "colortranslator";
 
+const customColors: {
+	[key: string]: string;
+} = {
+	transparent: "#00000000",
```
Comment (uncenter) [inline src/commands/utils/color.ts:9]: Just did `transparent` so far and I could make the if statement just check for that but this just puts it in place if we want to add another extra color in the future.

**Review by ryanccn — APPROVED** [review]

**Comment by uncenter** [comment]
the timing lmfao

**Pull request merged**